src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,110 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private T[] elementData;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkSize();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is invalid");
+        }
+        checkSize();
+        System.arraycopy(elementData,
+                index,
+                elementData,
+                index + 1,
+                size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIfIndexIsValid(index, size);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIfIndexIsValid(index, size);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIfIndexIsValid(index, size);
+        T element = (T) elementData[index];
+        System.arraycopy(elementData,
+                index + 1,
+                elementData,
+                index,
+                size - index - 1);
+        elementData[--size] = null;
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (elementData[i] == element
+                    || elementData[i] != null && elementData[i].equals(element)) {
+                return remove(i);
+            }
+        }
+
+        throw new NoSuchElementException("There's no such element");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIfIndexIsValid(int index, int length) {
+        if (index >= length || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is invalid");
+        }
+    }
+
+    private void checkSize() {
+        if (size == elementData.length) {
+            grow();
+        }
+    }
+
+    private void grow() {
+        T[] newElementData = (T[]) new Object[(size >> 1) + size];
```

Comment:
I hope you know what does (size >> 1) mean

Reply:
I know what that means. It is a bitwise offset. (size >> 1) means size / 2